modules/core/keeper/msg_server.go

```diff
@@ -122,13 +123,19 @@ func (k *Keeper) IBCSoftwareUpgrade(goCtx context.Context, msg *clienttypes.MsgI
 		return nil, errorsmod.Wrapf(ibcerrors.ErrUnauthorized, "expected %s, got %s", k.GetAuthority(), msg.Signer)
 	}
 
-	ctx := sdk.UnwrapSDKContext(goCtx)
-	upgradedClientState, err := clienttypes.UnpackClientState(msg.UpgradedClientState)
-	if err != nil {
-		return nil, errorsmod.Wrapf(clienttypes.ErrInvalidClientType, "cannot unpack client state: %s", err)
+	// TODO(sdk/v0.52): This should be reverted before release pending sdk fix. Accessing GetCachedValue returns nil due to loss of data in internal sdk msg handler.
+	// For now we can assume tendermint client state here as that's what we expect, but this may not always be the case for other msg handlers
+	// which use pb.Any encoding of msg types going through msg router service hybrid handler.
+	// upgradedClientState, err := clienttypes.UnpackClientState(msg.UpgradedClientState)
+	// if err != nil {
+	// 	return nil, errorsmod.Wrapf(clienttypes.ErrInvalidClientType, "cannot unpack client state: %s", err)
+	// }
+	var upgradedClientState ibctm.ClientState
+	if err := k.cdc.Unmarshal(msg.UpgradedClientState.Value, &upgradedClientState); err != nil {
+		return nil, errorsmod.Wrapf(clienttypes.ErrInvalidClientType, "cannot unmarshal tendermint client state: %s", err)
```

Comment:
https://github.com/cosmos/cosmos-sdk/issues/22779

I should add the link to the sdk issue inline in comments

Reply:
We should also have an issue in our repo (and put in the board) that links to this code to make sure we don't forget.

Reply:
Created an issue and put onto board https://github.com/cosmos/ibc-go/issues/7625